### DISCUSSION TITLE
Path to the js file with hash in generated html

### DISCUSSION THREAD

**Discussion by AndreyGudin** [webdiscus/pug-plugin#12] [General]
**[Question]**

Can pug-plugin resolve the path to the js file with name with hash and put it in to head of html file? Or does it needs another module, maybe babel?

**Comment by webdiscus** [comment]
Currently the plugin can't do that, but it's on the TODO list.
This feature is in high demand and later I'll do that.

There are so-called "asset manifest plugins" that replace normal filename with hashed version.

**Comment by AndreyGudin** [comment]
I understand, thank you

**Comment by webdiscus** [comment]
Hello @AndreyGudin,

I have good news, finally the resolving of scripts in pug is realized :-)

The new version `^2.0.0` supports the require() of styles and scripts directly in pug without having to define them in a webpack entry.

```pug
html
  head
    link(rel='stylesheet' href=require('./styles.scss'))
    script(src=require('./main.js'))
  body
```

Output:
```html
<!DOCTYPE html>
<html>
  <head>
    <link rel="stylesheet" href="/assets/css/styles.05e4dd86.css">
    <script src="/assets/js/main.f4b855d8.js"></script>
  </head>
  <body></body>
</html>
```

webpack.config.js
```js
const PugPlugin = require('pug-plugin');
module.exports = {
  output: {
    publicPath: '/',
    // js output filename 
    filename: 'assets/js/[name].[contenthash:8].js',
  },
  entry: {
    // all scripts and styles can be defined directly in pug
    'index': './src/index.pug',
  },
  plugins: [
    new PugPlugin({
      modules: [
        PugPlugin.extractCss({
          // css output filename 
          filename: 'assets/css/[name].[contenthash:8].css',
        }),
      ],
    }),
  ],
  module: {
    rules: [
      {
        test: /\.(pug)$/,
        loader: PugPlugin.loader,
        options: {
          method: 'render',
        }
      },
      {
        test: /\.scss$/i,
        use: ['css-loader', 'scss-loader'],
      },
    ],
  },
};
```

See the [test case](https://github.com/webdiscus/pug-plugin/tree/master/test/cases/require-scripts-render).

**Comment by webdiscus** [comment]
@AndreyGudin can you please test in your project whether it works for you?

**Comment by AndreyGudin** [comment]
Hi @webdiscus. It's working in my project. 
```
<!DOCTYPE html><html lang="ru"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><link href="style.css" rel="stylesheet"><script type="module" src="pages/main/scripts.js"></script>
```

```
module.exports={
  mode: mode,
  output:{
    filename:'pages/[name]/scripts.js',
    path:path.resolve(__dirname,'shelter'),
    publicPath:'',
    assetModuleFilename:'assets/[hash][ext][query]',
    clean: true,
  },
  entry:{
    'main/index':'./src/pages/main/main.pug',
    'pages/main/style':'./src/pages/main/main.scss',
    'our-pets/index':'./src/pages/our-pets/our-pets.pug',
    'pages/our-pets/style':'./src/pages/our-pets/our-pets.scss',
  },
  devServer: {
    open: true,
    watchFiles:['src/**'],
    static: {
      directory: path.resolve(__dirname, './shelter/pages'),
    }
  },
  devtool: 'source-map',
  plugins:[
    new PugPlugin({
      modules:[
        PugPlugin.extractCss({
          filename: '[name].css',
          postprocess:(content) =>{
            return content.replace(/assets/g,'../../assets');
          }
        }),
        { 
          test: /\.pug$/,
          outputPath:'shelter/pages/',
          postprocess:(content) =>{
            return content.replace(/assets/g,'../../assets');
          }
        }
      ]
    })
  ],
  module: {
    rules:[
      {
        test: /\.pug$/,
        loader: PugPlugin.loader,
        options:{
          method: 'render',
        },
      },
      {
        test: /\.scss$/,
        use:[
          "css-loader","sass-loader"
        ]
      },
      {
        test: /\.(png|svg|jpg|jpeg|gif)$/i,
        type:'asset/resource',
        exclude: path.resolve(__dirname, './src/fonts'),
        generator: {
          filename: 'assets/img/[hash][ext][query]'
        }
      },
      {
        test: /\.(woff|woff2|eot|ttf|otf|svg)$/i,
        type:'asset/resource',
        include: path.resolve(__dirname, './src/fonts'),
        generator: {
          filename: 'assets/fonts/[hash][ext][query]'
        }
      },
    ]
  },
};
```
![image](https://user-images.githubusercontent.com/43149000/161423350-e1216f89-4a29-4f69-b69b-dcf41165e5d3.png)


There is a problem, for example, if the js file is in the same directory as the html file, then the path becomes invalid. But anyway, this is great work.! Thanks!

**Comment by webdiscus** [comment]
Hello @AndreyGudin,

can you please create a simple repo with the reproducible problem and describe in readme what is expected as correct result.

1. instead of the filename replacing in `postprocess` you can use the option `filename` as function to modify the output filename:
```js
filename: (pathData, assetInfo) {
  if (something) {
     return 'some/path/to/filename'
  }
  pathData.filename; // original filename
}
```
The structure of `pathData` see [here](https://webpack.js.org/configuration/output/#template-strings).

2. the template substitution `[name]` for required styles/scripts in pug is the base name of required file, the `[name]` of the `./src/pages/about/script.js` is `script`.
 
Tipp, to use a specific output name:
- define specific base filename, e.g. `./src/pages/about/about.js`
- define a custom path in option `filename`, e.g.: `filename: 'pages/[name]/scripts.[contenthash:4].js'`
```pug
script(src=require(`./src/pages/about/about.js`))
```
output
```html
<script src="/pages/about/scripts.80da.js">
```

3. recommended use the absolute pablicPath, started with `/`, then an asset path in HTML will be always correct.

You can see the test case [options-outputpath](https://github.com/webdiscus/pug-plugin/tree/master/test/cases/options-outputpath). Here is the example how generate scripts and styles in same directory as the html file. In the `expected` directory see the generated files: html, js, css - all in same dir.

P.S. I have the idea to use the `[name]` as webpack entry name of issuer file instead of the base name of an asset.
```js
output: {
  path: path.join(__dirname, 'public/'),
  filename: 'pages/[name].js',
},
entry: {
  'about/index': './src/pages/about/main.pug'
},
plugins: [
  new PugPlugin({
    modules: [
      ...
      {
        test: /\.pug$/,
        outputPath: 'pages/', // relative by the webpack output.path
      },
    ],
  }),
],
```
./src/pages/about/main.pug
```pug
script(src=require(`./src/pages/about/scripts.js`))
```
then the output files: 
- `pages/about/index.html`
- `pages/about/index.js`

What do you think about this?
What should be the `[name]`: the base name of source file or the key-name of wepack entry?

**Comment by webdiscus** [comment]
Yet one thing..
I have fixed the option `outputPath` in version `2.0.1`. Now if the `outputPath` is relative then the output path is relative by webpack output.path. In older versions was supported only absolute path.

**Comment by AndreyGudin** [comment]
HI @webdiscus. Sorry for delay. Files index.html and script.js are in one folder, but path in head of index.html is:
```
src="pages/main/scripts.js"></script>
```
I expected
```
src="scripts.js"></script>
```

> What should be the [name]: the base name of source file or the key-name of wepack entry?

I think the key-name of webpack entry is fine, less confusion. By default, it is  the key-name of webpack entry in all plugins.

**Comment by webdiscus** [comment]
Hello @AndreyGudin 

> HI @webdiscus. Sorry for delay. Files index.html and script.js are in one folder, but path in head of index.html is:
> 
> ```
> src="pages/main/scripts.js"></script>
> ```
> 
> I expected
> 
> ```
> src="scripts.js"></script>
> ```

The pug-plugin do not affect on script filename. It is defined in `output. publicPath` and `output.filename`.
In index.html is the output filename of the script, generated by webpack.output options.

If the generated output filename is `pages/main/scripts.js` so will be in script src. Other is not possible ;-)
Please define `output.publicPath` as root `/`, then the filename will be correct works `<script src="/pages/main/scripts.js"></script>`



>  > What should be the [name]: the base name of source file or the key-name of wepack entry?
> 
> I think the key-name of webpack entry is fine, less confusion. By default, it is the key-name of webpack entry in all plugins.

That would "nice to have", but has many problems by realisation.

For example, we have one entry-point with the name `page/home`:
```js
entry: {
  'page/home': './src/pages/home/index.pug',
}
```

This pug file contains more scripts from different folder but with same filename.
Then output filename for all required scripts in the pug will be same: 
```pug
script(src=require(`./src/pages/home/index.js`)) -> page/home/index.js
script(src=require(`./src/scripts/libs/gallery/index.js`)) -> page/home/index.1.js (sufix `.1` is for an unique file)
```

It will be even more confusion than it is now.
Аnyway, in the output options is the function `filename(pathData)`.
The argument `pathData` contains path of the js source file and the path of the issuer (pug). 
You can use these paths to generate your own output filename.
So, currently this stay as is.
